### PR TITLE
SOL-58059 PauseResume

### PR DESCRIPTION
**Pull request by carolmorneau**
This implements the Pausable interface much like the older, active PR https://github.com/SolaceProducts/solace-spring-cloud/pull/57 

Please review/merge below PRs before this one as they were merged into this branch to facilitate testing: 
- https://github.com/SolaceProducts/solace-spring-cloud/pull/113
- https://github.com/SolaceProducts/solace-spring-cloud/pull/122

### CODE REVIEW

**Review by Nephery — CHANGES_REQUESTED** [review]

**Review by Nephery — CHANGES_REQUESTED** [review]
`JCSMPInboundChannelAdapter` doesn't persist its paused state between bindings being restarted (`STARTED` -> `PAUSED` -> `STOPPED` -> `STARTED`) for async consumers. Polled consumers work fine.

File: solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/JCSMPInboundChannelAdapter.java [hunk 1]
```diff
@@ -236,6 +241,28 @@ private InboundXMLMessageListener buildListener(FlowReceiverContainer flowReceiv
 		return listener;
 	}
 
+	@Override
+	public void pause() {
+		if (!isRunning()) {
+			logger.info(String.format("Can't pause consumer flows within inbound adapter %s as they are not running", id));
+			return;
+		}
```
Comment (Nephery) [inline solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/JCSMPInboundChannelAdapter.java:249]: Pause must work regardless of the consumer's running state.

File: solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/JCSMPInboundChannelAdapter.java [hunk 2]
```diff
@@ -236,6 +241,28 @@ private InboundXMLMessageListener buildListener(FlowReceiverContainer flowReceiv
...
+			fn.accept(true);
+		}
+	}
+
+	@Override
+	public void resume() {
+		if (!isRunning()) {
+			logger.info(String.format("Can't resume consumer flows within inbound adapter %s as they are not running", id));
+			return;
+		}
```
Comment (Nephery) [inline solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/JCSMPInboundChannelAdapter.java:260]: Resume must work regardless of consumer's running state.

File: solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/JCSMPMessageSource.java [hunk 3]
```diff
@@ -198,4 +201,45 @@ public void setPostStart(Consumer<Queue> postStart) {
...
+
+	@Override
+	public void pause() {
+		Lock writeLock = readWriteLock.writeLock();
+		writeLock.lock();
+		try {
+			if (!isRunning()) {
+				logger.info(String.format("Can't pause message source %s as it is not running", id));
+				return;
+			}
```
Comment (Nephery) [inline solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/JCSMPMessageSource.java:213]: Pause must work regardless of polled consumer's running state.

File: solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/JCSMPMessageSource.java [hunk 4]
```diff
@@ -198,4 +201,45 @@ public void setPostStart(Consumer<Queue> postStart) {
...
+
+	@Override
+	public void resume() {
+		Lock writeLock = readWriteLock.writeLock();
+		writeLock.lock();
+		try {
+			if (!isRunning()) {
+				logger.info(String.format("Can't resume message source %s as it is not running", id));
+				return;
+			}
```
Comment (Nephery) [inline solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/JCSMPMessageSource.java:231]: Resume must work regardless of polled consumer's running state.

**Review by Nephery — CHANGES_REQUESTED** [review]

**Review by Nephery — APPROVED** [review]
@carolmorneau I'm done with the PR fixes. Feel free to retest this when you have time.

File: solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/JCSMPMessageSource.java [hunk 5]
```diff
@@ -225,21 +229,33 @@ public void resume() {
...
 
 	@Override
 	public boolean isPaused() {
-		return flowReceiverContainer.isPaused();
+		if (paused) {
+			if (flowReceiverContainer.isPaused()) {
+				return true;
+			} else {
+				logger.warn(String.format(
+						"Flow receiver container %s is unexpectedly paused for message source %s",
```
Comment (carolmorneau) [inline solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/JCSMPMessageSource.java]: should this say: "Flow receiver container %s is unexpectedly **running** for message source %s "?

Comment (Nephery) [reply solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/JCSMPMessageSource.java]: Yup, it does. Look at the latest commit. I think you're looking at a diff that's one commit behind.

**Pull request merged**